### PR TITLE
docs/md/guides: database per tenant

### PR DESCRIPTION
**Pull request by rotemtam**

### CODE REVIEW

File: doc/md/guides/database-per-tenant/01-intro.mdx [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+providing better security and privacy guarantees.
+2. **Data Compliance**: By isolating tenants on the database level, you can ensure that each tenant's data is stored in
+compliance with data regulations and privacy laws relevant to that tenant. For example, you can host the specific
+database in a region that complies with the tenant's data residency requirements or even in your customer's own
+infrastructure.
+3. **Performance**: By isolating tenants on the database level, you can ensure that the performance of one tenant's
+database operations does not impact the performance of other tenants - eliminating the "noisy neighbor" problem.
+Additionally, you can utilize different instance types or configurations for each tenant's database to meet their
+specific performance requirements.
+4. **Simplified Development** (compared to shared database architectures) Codebases for multi-tenant applications are
```
Comment (hilakashai) [inline doc/md/guides/database-per-tenant/01-intro.mdx:43]:
```suggestion
4. **Simplified Development** (compared to shared database architectures): Codebases for multi-tenant applications are
```

File: doc/md/guides/database-per-tenant/01-intro.mdx [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+notoriously complex. Every action against the database must be made "tenant-aware" to ensure that data is correctly
+isolated. By isolating tenants on the database level, applications become effectively single-tenant, simplifying
+development, reducing the mental overhead of handling multi-tenancy, and reducing the risk of data leakage between tenants.
+5. **Simplified Deployment** (compared to full isolation) In a full isolation architecture, each tenant has its own
```
Comment (hilakashai) [inline doc/md/guides/database-per-tenant/01-intro.mdx:47]:
```suggestion
5. **Simplified Deployment** (compared to full isolation): In a full isolation architecture, each tenant has its own
```

File: doc/md/guides/database-per-tenant/01-intro.mdx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+deployment, requiring separate infrastructure, monitoring, and maintenance. By isolating tenants on the database level,
+you can share infrastructure, monitoring, and maintenance across tenants, reducing operational overhead.
+
+### Challenges
+
+> This architecture decisions is one of my biggest regrets, and we are currently in the process of rebuilding into a single database model.
```
Comment (hilakashai) [inline doc/md/guides/database-per-tenant/01-intro.mdx:53]:
```suggestion
> This architecture decision is one of my biggest regrets, and we are currently in the process of rebuilding into a single database model.
```

File: doc/md/guides/database-per-tenant/01-intro.mdx [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+>
+> *[HN Thread on Database per Tenant Architecture](https://news.ycombinator.com/item?id=23305111)*
+
+Despite their numerous advantages, Database per Tenant architectures present unique challenges, mostly around managing
+database schema migrations:
+
+* **Migration Duration** - As the number of tenants grows, deployment time increases linearly with the number of tenants.
+This is especially problematic when deploying schema migrations that take time, such as data migrations or index rebuilds.
+* **Schema Consistency** Ensuring that all tenant databases have consistent schema versions can be complex. Any schema
```
Comment (hilakashai) [inline doc/md/guides/database-per-tenant/01-intro.mdx:62]:
```suggestion
* **Schema Consistency** - Ensuring that all tenant databases have consistent schema versions can be complex. Any schema
```

File: doc/md/guides/database-per-tenant/01-intro.mdx [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+change needs to be applied across all tenant databases, which can be time-consuming and error-prone.
+* **Detecting Drift** - Migrations can fail due to various reasons, such as network issues, database downtime,
+or data dependent issues (such as constraint violations).  Such inconsistencies can cause application downtime for
```
Comment (hilakashai) [inline doc/md/guides/database-per-tenant/01-intro.mdx:65]:
```suggestion
or data dependent issues (such as constraint violations). Such inconsistencies can cause application downtime for
```

File: doc/md/guides/database-per-tenant/01-intro.mdx [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+users, as new application versions are rolled out and expect database changes to already have been applied.
+Detecting these inconsistencies (schema-drift) across a large fleet of databases can be challenging.
+* **Rollbacks** - Rolling back schema changes across a large fleet of databases can be challenging.  In a shared database
```
Comment (hilakashai) [inline doc/md/guides/database-per-tenant/01-intro.mdx:68]:
```suggestion
* **Rollbacks** - Rolling back schema changes across multiple databases can be difficult. In a shared database
```
You use the same sentence one above

File: doc/md/guides/database-per-tenant/01-intro.mdx [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+architecture, rolling back a schema change is a single operation. In a database per tenant architecture, rolling back
+a schema change requires rolling back the change on each tenant's database, which needs to be somehow orchestrated.
```
Comment (hilakashai) [inline doc/md/guides/database-per-tenant/01-intro.mdx:70]:
```suggestion
a schema change requires rolling back the change on each tenant's database, which needs to be orchestrated somehow.
```

File: doc/md/guides/database-per-tenant/02-db-groups.mdx [hunk 8]
```diff
@@ -0,0 +1,323 @@
+---
+title: Defining Target Groups
+id: target-groups
+slug: target-groups
+---
+In Atlas, a target group is a collection of target databases whose schema is managed together. In a database-per-tenant
+architecture, each tenant's database is a target database, and all tenant databases are grouped into a target group.
+However, you can also group databases by other criteria, such as environment (dev, staging, prod), region, or any other
+criteria that makes sense for your application.
```
Comment (hilakashai) [inline doc/md/guides/database-per-tenant/02-db-groups.mdx:9]:
```suggestion
conditions that makes sense for your application.
```

File: doc/md/guides/database-per-tenant/02-db-groups.mdx [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+
+For example, you might group all databases in the same region into a target group to ensure that schema changes are
+applied consistently across all databases in that region, or to group free-tier databases separately from
```
Comment (hilakashai) [inline doc/md/guides/database-per-tenant/02-db-groups.mdx:12]:
```suggestion
applied consistently across all databases in that region. Alternatively, you could group free-tier databases separately from
```

File: doc/md/guides/database-per-tenant/02-db-groups.mdx [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+paid-tier databases.
+
+Target groups can be defined statically or to be dynamically loaded from an API endpoint or a database query.
```
Comment (hilakashai) [inline doc/md/guides/database-per-tenant/02-db-groups.mdx:15]:
```suggestion
Target groups can be defined statically or dynamically loaded from an API endpoint or a database query.
```

File: doc/md/guides/database-per-tenant/02-db-groups.mdx [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+
+Target groups are defined in the project's `atlas.hcl` file and are later used by the Atlas CLI during the deployment
```
Comment (hilakashai) [inline doc/md/guides/database-per-tenant/02-db-groups.mdx:17]: link atlas HCL file

File: doc/md/guides/database-per-tenant/02-db-groups.mdx [hunk 12]
```diff
@@ -0,0 +1,323 @@
+---
+title: Defining Target Groups
+id: target-groups
+slug: target-groups
+---
+In Atlas, a target group is a collection of target databases whose schema is managed together. In a database-per-tenant
```
Comment (hilakashai) [inline doc/md/guides/database-per-tenant/02-db-groups.mdx:6]:
```suggestion
In Atlas, a **target group** is a collection of target databases whose schema is managed together. In a database-per-tenant
```

File: doc/md/guides/database-per-tenant/02-db-groups.mdx [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+:::info The `urlsetpath` function
+
+The `urlsetpath` function is a helper function provided by Atlas that allows you to set the "path" part
+of a URL. For example:
+
+```hcl
+urlsetpath("postgres://root:pass@localhost:5432", "mydb")
+#  ↳ Evaluates to "postgres://root:pass@localhost:5432/mydb"
+
+urlsetpath("mysql://localhost:3306", "mydb") # Evaluates to "mysql://localhost:3306/mydb"
```
Comment (hilakashai) [inline doc/md/guides/database-per-tenant/02-db-groups.mdx:94]: you have "evaluates to" twice here ?

File: doc/md/guides/database-per-tenant/02-db-groups.mdx [hunk 14]
```diff
@@ -0,0 +1,323 @@
...
+  tenants = local.decoded.tenants
+}
+
+env "tenants" {
+  for_each = toset(local.tenants)
+  url = urlsetpath("postgres://root:pass@localhost:5432", each.value)
+}
+```
+
+Next, we define an environment block for this target groups that consumes the `target_tenants` local variable
```
Comment (hilakashai) [inline doc/md/guides/database-per-tenant/02-db-groups.mdx:130]:
```suggestion
Next, we define an environment block for this target group that consumes the `target_tenants` local variable
```

File: doc/md/guides/database-per-tenant/02-db-groups.mdx [hunk 15]
```diff
@@ -0,0 +1,323 @@
...
+* We define a local variable `f` that reads the contents of the `tenants.json` file.
+* Next, we use the `jsondecode` function to parse the JSON content into a structured object.
+* We extract the `tenants` array from the decoded JSON object and store it in the `tenants` local variable.
+* Finally, we define an `env` block named `tenants` that iterates over the `tenants` list. For each tenant, we set the `url`
+  attribute to the result of the `urlsetpath` function, which combines the base URL with the tenant name.
+
+## Loading Data from an API Endpoint
+
+In some cases, you may want to load target groups dynamically from an API endpoint. For example, you might have a
+service `tenant-svc` that provides a list of tenant databases based on some criteria.  Let's suppose this service's
```
Comment (hilakashai) [inline doc/md/guides/database-per-tenant/02-db-groups.mdx:150]:
```suggestion
service `tenant-svc` that provides a list of tenant databases based on some criteria. Let's suppose this service's
```

File: doc/md/guides/database-per-tenant/02-db-groups.mdx [hunk 16]
```diff
@@ -0,0 +1,323 @@
...
+  "databases": [
+    "acme_corp",
+    "widget_inc",
+    "wayne_enterprises",
+    "stark_industries"
+  ]
+}
+```
+
+You can use the `runtimevar` data source to fetch with the `http` scheme to fetch this data and use it to define
```
Comment (hilakashai) [inline doc/md/guides/database-per-tenant/02-db-groups.mdx:165]:
```suggestion
You can use the `runtimevar` data source with the `http` scheme to fetch this data and use it to define
```

File: doc/md/guides/database-per-tenant/02-db-groups.mdx [hunk 17]
```diff
@@ -0,0 +1,323 @@
...
+* Next, we use the `jsondecode` function to parse the JSON content into a structured object.
+* We extract the `tenants` array from the decoded JSON object and store it in the `tenants` local variable.
+* Finally, we define an `env` block named `tenants` that iterates over the `tenants` list. For each tenant, we set the `url`
+  attribute to the result of the `urlsetpath` function, which combines the base URL with the tenant name.
+
+## Loading Data from an API Endpoint
+
+In some cases, you may want to load target groups dynamically from an API endpoint. For example, you might have a
+service `tenant-svc` that provides a list of tenant databases based on some criteria.  Let's suppose this service's
+endpoints recieve the target group id in the path, such as `https://tenant-svc/api/target-group/{id}` and return a
```
Comment (hilakashai) [inline doc/md/guides/database-per-tenant/02-db-groups.mdx:151]:
```suggestion
endpoints recieve the target group ID in the path, such as `https://tenant-svc/api/target-group/{id}` and return a
```

File: doc/md/guides/database-per-tenant/02-db-groups.mdx [hunk 18]
```diff
@@ -0,0 +1,323 @@
...
+
+## Incorporating Sensitive Data
+
+Essentially, defining target groups in Atlas is about dynamically compiling a list of URLs that represent the target
+databases. Database URLs often contain sensitive information, such as passwords, that should not be hardcoded in the
+`atlas.hcl` file, which is typically checked into version control.
+
+To address this issue, Atlas provides mechanisms for loading credentials from external sources, such as environment
+variables or secret management systems. This allows you to keep your database credentials secure while still being able
+to define target groups dynamically. Learn more about [Working with Secrets](/guides/deploying/secrets).
```
Comment (hilakashai) [inline doc/md/guides/database-per-tenant/02-db-groups.mdx:250]:
```suggestion
to define target groups dynamically. Learn more about [working with secrets](/guides/deploying/secrets).
```

File: doc/md/guides/database-per-tenant/02-db-groups.mdx [hunk 19]
```diff
@@ -0,0 +1,323 @@
...
+```hcl
+urluserinfo("postgres://localhost:5432", "root", "p455w0rd")
+#  ↳ Evaluates to "postgres://root:p455w0rd@localhost:5432"
+```
+
+:::
+
+## Next Steps
+
+In this guide, we've explored various techniques for defining target groups in Atlas. By using the `env` blocks and
```
Comment (hilakashai) [inline doc/md/guides/database-per-tenant/02-db-groups.mdx:318]:
```suggestion
In this guide, we've explored various techniques for defining target groups in Atlas. By using `env` blocks and
```

File: doc/md/guides/database-per-tenant/03-deploying.mdx [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+id: deploying
+slug: deploying
+---
+In the previous section, we learned how to define target groups in Atlas to manage migrations for a database-per-tenant
+architecture. In this section, we will see how to deploy migrations to the target groups.
+
+## Setting up
+
+For the purpose of this guide, we will use a simple example to demonstrate how to deploy migrations to
+target groups.  To simplify things, we will be using SQLite files as our target databases and statically
```
Comment (hilakashai) [inline doc/md/guides/database-per-tenant/03-deploying.mdx:12]:
```suggestion
target groups. To simplify things, we will be using SQLite files as our target databases and statically
```

File: doc/md/guides/database-per-tenant/04-cloud.mdx [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+```
+
+In the editor, add the following SQL statements:
+
+```sql
+INSERT INTO users (id, name) VALUES (1, "a8m");
+INSERT INTO users (id, name) VALUES (2, "rotemtam");
+```
+
+Save the file and exit the editor. Let's now push the new migration to the Schema Registry:
```
Comment (hilakashai) [inline doc/md/guides/database-per-tenant/04-cloud.mdx:100]:
```suggestion
Save the file and exit the editor. Let's push the new migration to the Schema Registry:
```

File: doc/md/guides/database-per-tenant/04-cloud.mdx [hunk 22]
```diff
@@ -0,0 +1,201 @@
...
+If an error occurs during a migration, having a centralized view of all your deployments can help you quickly
+identify the issue and take corrective action. You can view the error message, the target database, and the
+migration that caused the error.
+
+Suppose we run a deployment that fails during the schema migration phase, we can easily locate the error in the
+Atlas Cloud UI by navigating to the "Migrations" tab:
+
+![](https://atlasgo.io/uploads/db-per-tenant/migrations-screen.png)
+
+We easily locate the failed deployment and drill down to diagnose the issue:
```
Comment (hilakashai) [inline doc/md/guides/database-per-tenant/04-cloud.mdx:175]:
```suggestion
We quickly find the failed deployment and drill down to diagnose the issue:
```

File: doc/md/guides/database-per-tenant/04-cloud.mdx [hunk 23]
```diff
@@ -0,0 +1,201 @@
...
+on `tenant_4.db` with the error message:
+
+```
+Error: sql/migrate: executing statement "INSERT INTO users (id, name) VALUES (1, \"a8m\");" from version "20240721111345": UNIQUE constraint failed: users.id
+```
+We can further drill down into the specific database target migration:
+
+![](https://atlasgo.io/uploads/db-per-tenant/deployment-error.png)
+
+We now clearly see the issue, our data migration failed due to a unique constraint violation. We can now take corrective
```
Comment (hilakashai) [inline doc/md/guides/database-per-tenant/04-cloud.mdx:189]:
```suggestion
We now clearly see the issue, our data migration failed due to a unique constraint violation. Now, we can take corrective
```

File: doc/md/guides/database-per-tenant/04-cloud.mdx [hunk 24]
```diff
@@ -0,0 +1,201 @@
...
+## Conclusion
+
+In this section, we demonstrated how to use the Atlas Cloud Control Plane to manage migrations across multiple
+target databases. We showed how to push our project to the Atlas Cloud Schema Registry, deploy migrations to
+target databases, and gain visibility into the status of our deployments.
+
+While it is possible to manage migrations using the Atlas CLI, the Atlas Cloud Control Plane provides a centralized
+view of all your deployments, making it easier to manage and troubleshoot issues across multiple databases.
+
+
```
Comment (hilakashai) [inline doc/md/guides/database-per-tenant/04-cloud.mdx:201]:
```suggestion
```

File: doc/md/guides/database-per-tenant/01-intro.mdx (hunk 3 above)
Comment (rotemtam) [reply doc/md/guides/database-per-tenant/01-intro.mdx:53]: this is a quote.

File: doc/md/guides/database-per-tenant/02-db-groups.mdx (hunk 8 above)
Comment (rotemtam) [reply doc/md/guides/database-per-tenant/02-db-groups.mdx:9]: group by condition sounds weird

File: doc/md/guides/database-per-tenant/02-db-groups.mdx (hunk 13 above)
Comment (rotemtam) [reply doc/md/guides/database-per-tenant/02-db-groups.mdx:94]: 🤦

File: doc/md/guides/database-per-tenant/02-db-groups.mdx (hunk 8 above)
Comment (hilakashai) [reply doc/md/guides/database-per-tenant/02-db-groups.mdx:9]: So anything else cause you are using 'criteria' in the beginning of the sentence

File: doc/md/guides/database-per-tenant/02-db-groups.mdx (hunk 13 above)
Comment (rotemtam) [inline doc/md/guides/database-per-tenant/02-db-groups.mdx:94]:
```suggestion
urlsetpath("mysql://localhost:3306", "mydb")
```

File: doc/md/guides/database-per-tenant/01-intro.mdx [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+---
+
+## Intro
+
+This guide describes how to utilize Atlas to manage database schemas in "Database per Tenant" architectures,
+a common pattern for deploying multi-tenant applications.
+
+### What is a Database per Tenant Architecture?
+
+In a "Database per Tenant" architecture, each tenant has its own dedicated database. Database-per-tenant architectures are commonly
```
Comment (a8m) [inline doc/md/guides/database-per-tenant/01-intro.mdx:14]:
```suggestion
In a "Database per Tenant" architecture, each tenant has its own dedicated database (or a schema). Database-per-tenant architectures are commonly
```

File: doc/md/guides/database-per-tenant/02-db-groups.mdx [hunk 26]
```diff
@@ -0,0 +1,323 @@
...
+However, you can also group databases by other criteria, such as environment (dev, staging, prod), region, or any other
+criteria that makes sense for your application.
+
+For example, you might group all databases in the same region into a target group to ensure that schema changes are
+applied consistently across all databases in that region, or to group free-tier databases separately from
+paid-tier databases.
+
+Target groups can be defined statically or dynamically loaded from an API endpoint or a database query.
+
+Target groups are defined in the project's `atlas.hcl` file and are later used by the Atlas CLI during the deployment
```
Comment (rotemtam) [inline doc/md/guides/database-per-tenant/02-db-groups.mdx:17]:
```suggestion
Target groups are defined in the project's [`atlas.hcl`](/atlas-schema/projects) file and are later used by the Atlas CLI during the deployment
```

File: doc/md/guides/database-per-tenant/01-intro.mdx [hunk 27]
```diff
@@ -0,0 +1,79 @@
...
+Additionally, you can utilize different instance types or configurations for each tenant's database to meet their
+specific performance requirements.
+4. **Simplified Development** (compared to shared database architectures): Codebases for multi-tenant applications are
+notoriously complex. Every action against the database must be made "tenant-aware" to ensure that data is correctly
+isolated. By isolating tenants on the database level, applications become effectively single-tenant, simplifying
+development, reducing the mental overhead of handling multi-tenancy, and reducing the risk of data leakage between tenants.
+5. **Simplified Deployment** (compared to full isolation): In a full isolation architecture, each tenant has its own
+deployment, requiring separate infrastructure, monitoring, and maintenance. By isolating tenants on the database level,
+you can share infrastructure, monitoring, and maintenance across tenants, reducing operational overhead.
+
```
Comment (a8m) [inline doc/md/guides/database-per-tenant/01-intro.mdx:50]:
```suggestion
6. **Simplified Maintenance**: Running tenants on different databases simplifies maintenance tasks: Deleting, exporting, and loading tenant data do not require special toolings as most databases offer these functionalities.  
```

File: doc/md/guides/database-per-tenant/02-db-groups.mdx [hunk 28]
```diff
@@ -0,0 +1,323 @@
...
+    for_each = toset(local.target_db_urls)
+    url = each.value
+}
+```
+
+When the `for_each` meta-argument is used, the `env` block is instantiated for each value in the list, and the `each`
+object is used to access the current value. In our case, we will get two instances of the `target` block, one for each
+URL in the `target_db_urls` list.
+
+## Dynamically Compiling URLs
```
Comment (a8m) [inline doc/md/guides/database-per-tenant/02-db-groups.mdx:57]:
```suggestion
## Dynamically Computing URLs
```

File: doc/md/guides/database-per-tenant/03-deploying.mdx [hunk 29]
```diff
@@ -0,0 +1,144 @@
...
+In the previous section, we learned how to define target groups in Atlas to manage migrations for a database-per-tenant
+architecture. In this section, we will see how to deploy migrations to the target groups.
+
+## Setting up
+
+For the purpose of this guide, we will use a simple example to demonstrate how to deploy migrations to
+target groups. To simplify things, we will be using SQLite files as our target databases and statically
+defining the target groups in the `atlas.hcl` file.
+
+### Our project file
```
Comment (a8m) [inline doc/md/guides/database-per-tenant/03-deploying.mdx:15]:
```suggestion
### Our config file
```

**Review by a8m — APPROVED** [review]

**Pull request merged**